topology_state_load: stop waiting for IP-s

gusev-p opened the pull request:
The loop in `id2ip` lambda makes problems if we are applying an old raft log that contains long-gone nodes. In this case, we may never receive the `IP` for a node and stuck in the loop forever. In this series we replace the loop with an if - we just don't update the `host_id <-> ip` mapping in the `token_metadata.topology` if we don't have an `IP` yet.

The PR moves `host_id -> IP` resolution to the data plane, now it happens each time the IP-based methods of `erm` are called. We need this because IPs may not be known at the time the erm is built. The overhead of `raft_address_map` lookup is added to each data plane request, but it should be negligible. In this PR `erm/resolve_endpoints` continues to treat missing IP for `host_id` as `internal_error`, but we plan to relax this in the follow-up (see this PR first comment).

gusev-p commented:
discuss/follow-up:
  * in raft-topology mode in `replace-with-same-ip` scenario we call `storage_service::remove_endpoint` method for the left node, which has the same IP as the new replacing node. The `gossiper::quarantine_endpoint` will be called for this IP, which means all the gossiper events will be ignored for this IP for the duration of quarantine_delay=1 minute.
    * This problem existed before this patch
    * There is no such problem for gossiper mode, in `storage_service::handle_state_normal` we call `storage_service::remove_endpoint` only for IPs which don't own tokens, and in `replace-with-same-ip` scenario the IP will have owned tokens.
  *  In [storage_service: wait_for_ip for transition_nodes](https://github.com/scylladb/scylladb/pull/16639/commits/247ab4f534c0590181baea97fd691db4a710ce20) we call `wait_for_ip` for `transition_nodes`, which should include joining nodes and replacing nodes. This is not a bullet-proof solution though - a node might restart after receiving this barrier command, but before persisting it into system.peers table in topology_state_load. It will lose the mapping in this case.
  * We don't handle the case when a node changes its IP address. There is no raft event when the node is restarted after IP change -> no topology_state_load -> erm is not updated. There is a `topology/test_change_ip.py` test, but it doesn't check the data plane actually. We plan to switch erm to `raft_address_map` in the follow-up. This way IP changes won't require erm to be rebuilt.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5548/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6960/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5548/)
#### Failed Tests (30/23683):
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_raft_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_raft_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_raft_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_replacing_alive_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5548/testReport/junit/%28root%29/test_replace_alive_node/Tests___Unit_Tests___test_replacing_alive_node_fails) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replacing_alive_node_fails)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider1.cloudius-systems.com

kbr-scylla commented:
> In this series we replace the loop with an if - we just don't update the host_id <-> ip mapping in the token_metadata.topology if we don't have an IP yet.

And how is e_r_m calculated then?

gusev-p commented:
> And how is e_r_m calculated then?

Please be more specific.

kbr-scylla commented:
If token_metadata + replication strategy indicates that X is a replica for a given token, but we don't have IP for X. What will be the replica set for this token in e_r_m given that replica sets in e_r_m are represented as sets of IPs and uses topology to calculate the mappings

kbr-scylla commented:
(The answer to my question is probably in the commits, I just looked at the cover letter and immediately had this question)

gusev-p commented:
> If token_metadata + replication strategy indicates that X is a replica for a given token, but we don't have IP for X. What will be the replica set for this token in e_r_m given that replica sets in e_r_m are represented as sets of IPs and uses topology to calculate the mappings

A stub is used instead of real IP (`inet_address{}`), should be treated like any other unavailable node.

kbr-scylla commented:
> should be treated like any other unavailable node

Could you please perform an experiment and check what storage_proxy would do for such missing IP?
I mean for an actual node present in the cluster -- not some old / left node.

So for example using error injection, something like:
1. start 3 node cluster
2. create RF=3 table
3. stop one node
4. restart it with an error injection enabled so it pretends that we don't have IP address of one of the other nodes inside topology_state_load
5. perform a write using this node as coordinator, check e.g. that QUORUM write succeeds but ALL write fails (and what the error message will be)

gusev-p commented:
> > should be treated like any other unavailable node
> 
> Could you please perform an experiment and check what storage_proxy would do for such missing IP? I mean for an actual node present in the cluster -- not some old / left node.
> 
> So for example using error injection, something like:
> 
> 1. start 3 node cluster
> 2. create RF=3 table
> 3. stop one node
> 4. restart it with an error injection enabled so it pretends that we don't have IP address of one of the other nodes inside topology_state_load
> 5. perform a write using this node as coordinator, check e.g. that QUORUM write succeeds but ALL write fails (and what the error message will be)

OK

kbr-scylla commented:
> We plan to switch erm to raft_address_map in the follow-up. This way IP changes won't require erm to be rebuilt.

But if we switched erm to use raft_address_map underneath I understand it would still be doing IP translations inside when it's being built so we would still need to rebuild it in case of IP change / missing IP appearing.

Perhaps you mean the "follow-up to follow-up" where we switch erm to use host IDs only and raft_address_map is accessed on the messaging level?

gusev-p commented:
> > We plan to switch erm to raft_address_map in the follow-up. This way IP changes won't require erm to be rebuilt.
> 
> But if we switched erm to use raft_address_map underneath I understand it would still be doing IP translations inside when it's being built so we would still need to rebuild it in case of IP change / missing IP appearing.

My understanding is that the only consumer of `host_id<->IP` mapping from `token_metadata.topology` in raft topology mode is the `abstract_replication_strategy.cc/resolve_endpoints` function, if we switch it to `raft_address_map` we don't have to maintain the `token_metadata.topology` mapping, no?

kbr-scylla commented:
But the problem is that at the moment of calling `resolve_endpoints` -- when we're building e_r_m -- IP mapping might be missing, doesn't matter if we try to get from raft_address_map or topology, it might be missing everywhere, also from raft_address_map. Or it might be outdated, or even if it's up-to-date it might change after we built the e_r_m. So we will still have to rebuild e_r_m whenever IPs change.

kbr-scylla commented:
But it's true that the locator::topology mapping should become redundant.

The part I disagree with is
> IP changes won't require erm to be rebuilt.

gusev-p commented:
> But the problem is that at the moment of calling resolve_endpoints -- when we're building e_r_m

It's not the case, this PR changes this. Now `resolve_endpoints` is called from the data plane, and the erm data structures are `host_id`-based.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1992,8 +1992,12 @@ class topology_coordinator {
+                    raft_topology_cmd cmd{raft_topology_cmd::command::barrier};
+                    for (const auto& [id, _]: _topo_sm._topology.transition_nodes) {
+                        cmd.wait_for_ip.push_back(id);
+                    }
```
Why do you need to pass it explicitly. The receiving node can look at transition_nodes by itself.

gusev-p replied:
I wanted to have the IP mapping inplace before `read_barrier`/`topology_state_load`, so that `topology_state_load` could add the proper mapping into erm. Although on the other hand now I'm moving erm to `raft_address_map` and this requirement is no longer relevant.

gusev-p replied:
rewrote without explicit parameter, use `_topology_state_machine._topology.new_nodes`

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1992,8 +1992,12 @@ class topology_coordinator {
+                    }
                     auto f = co_await coroutine::as_future(exec_global_command(std::move(guard),
-                        raft_topology_cmd::command::barrier,
+                        cmd,
```
replace does not pass through commit_cdc_generation, but it should be done for replace as well, no?

gusev-p replied:
> replace does not pass through commit_cdc_generation

hmm, I hoped for that. Should we put the similar barrier for replace?

gleb-cloudius replied:
Replace goes to `topology::transition_state::tablet_draining` and there is no guaranty this state will call a barrier (only if there are tablets to drain). I guess we can add explicit barrier call there.

kbr-scylla replied:
Can't we do this `wait_for_ip` barrier inside `topology::transition_state::join_group0` which should always be the first step in bootstrap/replace?

gusev-p replied:
Added `wait_for_ip` command. We need to call it from `handle_node_transition`, before moving topology to transition, otherwise `topology_state_load` could observe ids without ips

kbr-scylla commented:
> It's not the case, this PR changes this. Now resolve_endpoints is called from the data plane, and the erm data structures are host_id-based.

Ah, this is major part of the PR, I'd say it's more important part than anything else. And it has implications for performance for example (doing the translations on the hot path). Can you include this info in cover letter? Isn't this basically the crux of the PR and even the title may deserve to be something like "transition e_r_m to host IDs"?

It's also a further change than you planned before AFAIR, when you wanted to keep e_r_m IP-based and rebuild it on IP changes.

gusev-p commented:
> Ah, this is major part of the PR, I'd say it's more important part than anything else.

This is subjective - which part is more important. To me, the ultimate goal is to remove the `id2ip` loop.

> And it has implications for performance for example (doing the translations on the hot path).

I remember mentioning exactly this point on one of out weekly-s with Avi, he was ok with that.

> Can you include this info in cover letter?

OK.

> Isn't this basically the crux of the PR and even the title may deserve to be something like "transition e_r_m to host IDs"?

This was the simplest part, and the goal is to solve the problems with `topology_state_load` for old log entries.

> It's also a further change than you planned before AFAIR, when you wanted to keep e_r_m IP-based and rebuild it on IP changes.

No, I wanted to switch erm to `host_id` from the beginning.

kbr-scylla commented on repair/repair.hh:
```diff
@@ -178,10 +178,10 @@ struct get_sync_boundary_response {
 using get_combined_row_hash_response = repair_hash;
```
> The problem is that repair_meta is addressed by IP. In replace-with-same-ip scenario it's possible that the new node bootstrap/repair process is started before the gossiper events reach some of the nodes, and then later when these events eventually come, they remove the new repair_meta, which is used by the ongoing repair process.


Why weren't we hitting this before? Is it because the wait in topology_state_load prevented repair from being started until these events reached everyone?

kbr-scylla commented on repair/row_level.cc:
```diff
@@ -2265,12 +2265,20 @@ future<repair_flush_hints_batchlog_response> repair_service::repair_flush_hints_
+static locator::host_id get_host_id(const rpc::client_info& cinfo) {
+    const auto* result = cinfo.retrieve_auxiliary_opt<locator::host_id>("host_id");
+    if (!result) {
+        throw std::runtime_error("host_id not found");
```
Can you add more context to the message?
Useful info may be:
- call site function name (by adding `seastar::compat::source_location` parameter with `current` default value to `get_host_id`, and using it in the message)
- IP address of `cinfo`

gusev-p replied:
dropped the entire commit

kbr-scylla commented on repair/row_level.cc:
```diff
@@ -3016,10 +3025,11 @@ class row_level_repair_gossip_helper : public gms::i_endpoint_state_change_subsc
     {}
     future<> remove_row_level_repair(gms::inet_address node) {
-        rlogger.debug("Started to remove row level repair on all shards for node {}", node);
+        const auto id = _repair_service.get_gossiper().get_host_id(node);
```
Given the purpose of the commit message is not to depend on gossiper propagation to be quick enough -- doesn't this line kinda defeat the purpose?

kbr-scylla replied:
So the scenario was:
- we replace-with-same-ip, old ID = X new ID = Y, IP = .1 let's say
- we start repair and create repair_meta for .1
- now for some reason we get gossiper notification with ID = X IP = .1
- we remove repair meta for IP .1 which is wrong

And after your changes we have:
- start repair and create repair_meta for ID = Y
- for some reason we get gossiper notification with ID = X IP = .1
- we remove repair meta for ID = X i.e. we don't remove anything because there is no repair meta

Correct?

kbr-scylla replied:
So in this scenario `get_host_id(.1)` will return `X` (old ID)

gusev-p replied:
It seems we no longer need this commit, `wait_for_ip` ensures roughly the same invariants as they were before this PR. I dropped this commit for now.

kbr-scylla commented on idl/storage_service.idl.hh:
```diff
@@ -32,6 +32,7 @@ struct raft_topology_cmd {
         fence
```
Maybe https://github.com/scylladb/scylladb/issues/16644 is the issue.

After our discussion with @gleb-cloudius -- one approach that we both like is to not update system.peers inside topology_state_load, but inside a separate subscriber. (That subscribes to both gossiper and topology_state_load calls.)

This could be the same subscriber which updates raft_address_map. And the raft_address_map update would happen after system.peers persisting. Therefore, `wait_for_ip` finishing would guarantee that the entry was persisted.

gusev-p replied:
mentioned this in the [todo list](https://github.com/scylladb/scylladb/pull/16639#issuecomment-1876567388)

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -435,36 +422,41 @@ future<> storage_service::topology_state_load() {
+                co_await _gossiper.add_local_application_state({{ gms::application_state::STATUS, gms::versioned_value::normal(rs.ring.value().tokens) }});
+            } else if (ip && !is_me(*ip)) {
+                // In replace-with-same-ip scenario the replaced node IP will be the same
+                // as ours, we shouldn't put it into system.peers.
```
If this is indeed an old Raft log entry that we're applying, which comes from a time where a replace-with-same-IP was happening, then this check might not always detect the case -- if `am.find` fails, right?

Maybe there's a more reliable way to detect the replace-with-same-IP scenario, from topology state or sth.

kbr-scylla replied:
> If am.find fails, we won't pass the condition on ip being present in this if.

That's my point and my question is whether it's a problem. We won't enter this `else if` branch, meaning we won't detect the replace-with-same-ip scenario, meaning we will execute the wrong logic for replace-with-same-ip.

Every time we enter this `else if`, we are guaranteed to be in replace-with-same-ip. But not every time we're in replace-with-same-ip, we will enter this `else if` -- because `am.find` is unreliable (it depends on timing). So we will not call `co_await _sys_ks.local().update_peer_info(*ip, info)` sometimes during replace-with-same-ip.

gusev-p replied:
We can only be the `replacing` node, the `being_replaced` node with this IP should be stopped and its state is not important.

The condition `if (is_me(host_id))` being false means we are considering some other normal node. The missing ip means we are postponing the local system table update until later times. This can be resolved either by topology coordinator, when it calls `wait_for_ip` for the actual new nodes (not some old irrelevant ones), of by `gossiper_state_change_subscriber_proxy` in case the `wait_for_ip` failed.

If the ip is known and it's equal to our IP, then we don't need to update `system.peers`. We should wait until the `being_replaced` become left, and then it will be removed from system.peers by the relevant code in `process_left_node`.

gusev-p commented:
new version:
  * introduced a new raft command `wait_for_ip`, invoke it from `handle_node_transition`;
  * reverted the `resolve_endpoints` change, it continue to threat missing IPs as `internal_error`-s for the sake of reducing the scope of this PR (left for the follow-up);
  * invoke `system.peers`, `token_metadata.topology` and `erm` synchronisation logic when `host_id<->IP` mapping changes in response to gossiper event;
  * improved `test_change_ip.py`.

[all the changes](https://github.com/scylladb/scylladb/compare/d6f1291403da7181d509a07a9aae2f438c087ab0..01f1f98d5b4aef4cba1fc17cc493650d220f99e0).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5621/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7021/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5621/)
#### Failed Tests (120/23683):
- [topology.test_cluster_features.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.1)
- [topology.test_cluster_features.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.2)
- [topology.test_cluster_features.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.3)
- [topology_experimental_raft.test_raft_cluster_features.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_cluster_features_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_cluster_features.debug.1)
- [topology_experimental_raft.test_raft_cluster_features.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_cluster_features_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_cluster_features.debug.2)
- [topology_experimental_raft.test_raft_cluster_features.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_cluster_features_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_cluster_features.debug.3)
- [topology_experimental_raft.test_topology_ops.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.debug.1)
- [topology_experimental_raft.test_topology_ops.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.debug.2)
- [topology_experimental_raft.test_topology_ops.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.debug.3)
- [topology.test_cluster_features.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5621/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.dev.1)
#### Build Details:
- Duration: 4 hr 51 min
- Builder: i-0a519dfca9d236436 (m5ad.8xlarge)

gusev-p commented:
new version:
  * tests fix: wait for ip before moving a node to left state and banning it

[the changes](https://github.com/scylladb/scylladb/compare/f82f8babb163eb9d707b97df414d0583c1ba6db7..8e8070df141e902c1101fb1354dd1b8e0134339d)

kbr-scylla commented:
> added try/catch https://github.com/scylladb/scylladb/compare/8e8070df141e902c1101fb1354dd1b8e0134339d..c83b4280253b2521f062f0551964a58230aabfb3

So we no longer, as the comment above says, "Make sure the IP is known" -- there is no longer any guarantee, so we might as well not do the `co_await wait_for_ip` at all. Why can it fail and why the try/catch?

kbr-scylla commented on idl/storage_service.idl.hh:
```diff
@@ -29,7 +29,8 @@ struct raft_topology_cmd {
         barrier,
```
> We dont't treat wait_for_ip errors as fatal, we just write a warning into the log if the command failed. The gossiper events might have been delayed for various reasons,

That's true but isn't that precisely why we wait? The gossiper event with IP must reach us eventually. Did you observe in tests that it never arrives?

kbr-scylla replied:
Ah, I think I get it -- this is for the case of bootstrap failing, therefore the gossiper event may never reach us. We might not be able to contact the node that failed to boot.

But the comment should elaborate on why the try/catch is needed.

In fact, why not put the `wait_for_ip` from beginning of `case node_state::none` into the case we want to reject the node? The case where we don't reject should be handled by the global barrier below, IIUC.

kbr-scylla replied:
I also don't understand why we can continue if wait_for_ip fails. We don't have the IP, we cannot communicate with the node -- how can we proceed?

gusev-p replied:
> Ah, I think I get it -- this is for the case of bootstrap failing, therefore the gossiper event may never reach us. We might not be able to contact the node that failed to boot.

Yes, but not only. The main reason is in the commit message:
> the missing IP for some node shouldn't be fatal for data plane
> request, this node should be treated just like any other
> unavailable node.

> In fact, why not put the wait_for_ip from beginning of case node_state::none into the case we want to reject the node? The case where we don't reject should be handled by the global barrier below, IIUC.

Why? Now the code is more straightforward for me. But we need to add `get_my_id` to the `exclude_nodes`...

kbr-scylla replied:
I don't understand what data plane has anything to do with the issue. This is control plane code.

kbr-scylla replied:
Does this control plane code have to communicate with the node?
If yes, we need the IP. I don't understand how we can skip getting the IP if we need to communicate with it.
If no, then we don't need the wait_for_ip at all.

kbr-scylla replied:
I understand that this part
```cpp
                        node.guard = co_await exec_global_command(std::move(node.guard),
                            raft_topology_cmd::command::wait_for_ip,
                            {node.id});
```

is necessary for the data plane to function.

But I also don't understand why you made it optional. We should fail bootstrap if we cannot learn the bootstrapping node's IP.

gleb-cloudius replied:
It is a philosophical question. If you proceed here and the IP will not be resolved before it is needed you will fail later. You already have (or suppose to have) code that handles the failure, so you may as well proceed here.

gusev-p replied:
> I don't understand what data plane has anything to do with the issue. This is control plane code.

The moment where we absolutely need an IP is when a node is coordinating a data plane request and it needs to direct it to another node. This is where we want to treat a missing IP for a node the same way as the node just being unavailable. This is why we have `try/catch` around `wait_for_ip` cmd call.

gusev-p replied:
> Does this control plane code have to communicate with the node?
> If yes, we need the IP. I don't understand how we can skip getting the IP if we need to communicate with it further down the line.
> If no, then we don't need the wait_for_ip at all.

We need an IP to deliver `join_node_response_params::rejected` to the node. We need to `wait_for_ip` before `update_topology_state(state := left)` since otherwise the node will be banned and we'll never get an IP. We need try/catch around `wait_for_ip(node.id)` because we need `update_topology_state(state := left)` for it event if we can't resolve an IP.

gusev-p replied:
> But I also don't understand why you made it optional. We should fail bootstrap if we cannot learn the bootstrapping node's IP.

Why fail bootstrap if some of the nodes can't resolve new node's IP? The other nodes can still work with it. And even if node X can't resolve ip of node Y it still can coordinate requests, just treating Y as unavailable.

kbr-scylla replied:
> Why fail bootstrap if some of the nodes can't resolve new node's IP? The other nodes can still work with it. And even if node X can't resolve ip of node Y it still can coordinate requests, just treating Y as unavailable.

By making the `wait_for_ip` optional you're reducing data path availability. For example temporarily CL=ALL requests might be failing during bootstrap, because we proceed without waiting for all nodes to learn about the node's IP. This is not a correct thing to do. Topology changes should not be introducing unavailability if there are no failures.

And if they cannot learn the IP eventually, it means there's some larger problem. If they can't learn it now, during boot, when the node is alive and gossiping with the cluster, why would it be a good idea to proceed with bootstrap? Any reason they will be able to learn the IP of this node when it finishes bootstrapping, if they can't learn it now? You'll just reduce availability even further once the node becomes NORMAL. So it's logical to fail bootstrap in this case.

kbr-scylla replied:
Above I was talking about the global `wait_for_ip`, the purpose of which is to support the data plane.

The other `wait_for_ip`, at the beginning of the transition, is used to support the control plane. And there it makes sense to make it optional, because delivering `join_node_response_params::rejected` is optional. But it needs a comment. And because it's used for the singular purpose of delivering `join_node_response_params::rejected`, I think it would be clearer if the wait was inside the branch that delivers `join_node_response_params::rejected`.

gusev-p replied:
made it non-optional

gusev-p commented:
> So we no longer, as the comment above says, "Make sure the IP is known" -- there is no longer any guarantee, so we might as well not do the `co_await wait_for_ip` at all. Why can it fail and why the try/catch?

Of course not, the IP will be there in most cases. The `try/catch` here is needed to match the corresponding `try/catch`-es around `wait_for_ip` cmd call and `try/catch` around `respond_to_joining_node`. The reason for `try/catch` in `wait_for_ip` cmd is outlined in the commit message:
> We dont't treat wait_for_ip errors as fatal, we just write
> a warning into the log if the command failed. The gossiper
> events might have been delayed for various reasons, the
> missing IP for some node shouldn't be fatal for data plane
> request, this node should be treated just like any other
> unavailable node. This is part is not implemented
> yet though and left for the follow-ups. For now erms continue
> to treat missing IPs as internal_error as it was
> before the patch.
>

kbr-scylla commented:
Yeah I don't get the commit message logic. If gossiper events weren't delayed we wouldn't need wait at all because we would always have IP immediately. The reason why we need the wait is precisely because we don't know how long it will take for gossiper event to reach us. But my point is it should reach us **eventually** so the call should always succeed.

Unless for example we assume that the node that we're mapping might be gone before gossiping with others.

In that case I agree, but this is only for specific scenarios (e.g. we're waiting for IP of bootstrapping node and boot may fail), not all cases.

Let's move the discussion to this  thread https://github.com/scylladb/scylladb/pull/16639#discussion_r1445955056

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5640/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7042/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5640/)
#### Failed Tests (62/23701):
- [test_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_snapshot/Tests___Unit_Tests___test_snapshot) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_snapshot)
- [test_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_snapshot/Tests___Unit_Tests___test_snapshot_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_snapshot)
- [test_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_snapshot/Tests___Unit_Tests___test_snapshot_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_snapshot)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_decommission_node_add_column](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_topology_remove_decom/Tests___Unit_Tests___test_decommission_node_add_column) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_node_add_column)
- [test_topology_schema_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_topology_schema/Tests___Unit_Tests___test_topology_schema_changes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_schema_changes)
- [test_removing_alive_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_remove_alive_node/Tests___Unit_Tests___test_removing_alive_node_fails) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removing_alive_node_fails)
- [test_nodes_with_different_smp](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5640/testReport/junit/%28root%29/test_topology_smp/Tests___Unit_Tests___test_nodes_with_different_smp) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nodes_with_different_smp)
#### Build Details:
- Duration: 2 hr 18 min
- Builder: spider4.cloudius-systems.com

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -614,6 +607,9 @@ future<> storage_service::topology_state_load() {
             continue;
         }
         const auto ep = tmptr->get_endpoint_for_host_id(e);
```
Yes. It looks like we skip something cdc related here.

gleb-cloudius replied:
> > Yes. It looks like we skip something cdc related here.
> 
> `handle_cdc_generation` doesn't care about IPs, we don't need it in `gossiper_state_change_subscriber_proxy`.

But we skip it if the mapping is not available in topology load.

gusev-p replied:
> But we skip it if the mapping is not available in topology load.

I don't understand.

In `topology_state_load` we have
```
if (auto gen_id = _topology_state_machine._topology.current_cdc_generation_id) {
        slogger.debug("topology_state_load: current CDC generation ID: {}", *gen_id);
        co_await _cdc_gens.local().handle_cdc_generation(*gen_id);
    }
```
it doesn't need the mappings.

gleb-cloudius replied:
Ah, right. It is outside of the loop, so it is not skipped.

kbr-scylla replied:
> there should be no regression since we wait for IPs before bootstrapping a node

Which you made optional (which I believe is incorrect) so the argument is invalid.

gusev-p replied:
made it non-optional

gusev-p commented on test/topology/test_change_ip.py:
```diff
@@ -27,5 +33,28 @@ async def test_change_two(manager, random_tables):
+        if not remaining:
+            return True
+        logger.info(f"waiting for {host0} to see its peers, all_ips {all_ips}, peers {peers}, remaining {remaining}")
+    await wait_for(see_peers, time.time() + 60)
```
this test exercises IP change scenario, we want to be sure that:
  * server sees the proper IPs (not the old ones)
  * the IPs are persisted in `system.local`

`manager.server_sees_others` doesn't provide that

gusev-p commented on service/storage_service.hh:
```diff
@@ -758,6 +763,10 @@ private:
     } _raft_topology_cmd_handler_state;
+    class gossiper_state_change_subscriber_proxy;
+    // A proxy class representing subscription to on_change
+    // events, and updating the address map on this events.
```
OK, will rename it in a separate commit

gusev-p commented:
new version:
  * dropped the `node_repair_meta_id -> host_id`, it's no longer needed;
  * fixed the deadlock in `gossiper_state_change_subscriber_proxy`, `read_apply_mutex` and gossiper locks could me acquired in cross-order, more thorough explanation is in the code commen
  * renamed `gossiper_state_change_subscriber_proxy -> raft_ip_address_updater`

[all the changes](https://github.com/scylladb/scylladb/compare/c83b4280253b2521f062f0551964a58230aabfb3..788809ee67f4b02839fe2dfde1f5b9fe9feb1e10)

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -7651,7 +7651,7 @@ future<> storage_service::wait_for_normal_state_handled_on_boot() {
 future<bool> storage_service::is_cleanup_allowed(sstring keyspace) {
     return container().invoke_on(0, [keyspace = std::move(keyspace)] (storage_service& ss) {
-        auto my_address = ss.get_broadcast_address();
+        auto my_address = ss.get_token_metadata().get_my_id();
```
The variable name should be updated too.

gusev-p commented:
> Isn't it a good idea to do it anyway? Perhaps in another PR.

Why bother with it, it was not a problem before, it shouldn't be a problem in the future either if `wait_for_ip` waits for all nodes. We can reconsider if we have a clear reproducer.

yaronkaikov commented:
@kbr-scylla @gusev-p Don't know why, but now we keep failing in gating (x86, debug).

gusev-p commented:
new version:
  * made `wait_for_ip` non-optional

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5737/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5737/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7113/)
#### Failed Tests (2/23715):
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5737/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [topology.test_change_ip.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5737/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_change_ip_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_change_ip.debug.1)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider3.cloudius-systems.com

gusev-p commented:
new version:
  * `test_change_two`, fix flakiness, add `driver_close/driver_connect`

kbr-scylla commented:
Qd!
